### PR TITLE
User Interface Tab for showing Terraform Plan

### PR DESCRIPTION
**Pull request by zanawar**
Solves #200 
This PR contains the UI Tab that display the Terraform Plan

### CODE REVIEW

File: src/plan-summary-tab/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+You can locally develop and test the summary tab.
+
+From the root project directory:
+
+1. Run `npm i` to install required packages
+2. Run `npm run compile:test` to compile the tab with test data
+3. If using VSCode you can use the Run/Debug Launch command `debug:test - summary tab - Chrome`
+    - Alternatively you can open the `./dist/plan-summary-tab/plan-summary-tab.html` file in a browser of your choice
+    - For breakpoints to be enabled you will need to launch Chrome with the `--remote-debugging-port=9222` command line option.
```
Comment (charleszipp) [inline src/plan-summary-tab/README.md:21]: This made me think I needed to update something manually in launch.json when testing this. However, debugger is attached as you have it configured. Would suggest to reword this more as an FYI. For example

```
Debugger will use the default remote debugging port of 9222. 
```

File: package.json [hunk 2]
```diff
@@ -3,12 +3,15 @@
   "version": "1.0.0",
   "description": "This contains the Azure Pipelines tasks for installing and executing terraform commands.",
   "scripts": {
+    "postinstall": "cd ./tasks/terraform-installer && npm i && cd ../terraform-cli && npm i && cd ../../",
```
Comment (charleszipp) [inline package.json:6]: I did not foresee that we would need to modify the root package.json but makes complete sense why you had to do this. This is making me think it might be best to go ahead and move this into its own directory with its own package.json. This mainly so the dependencies for packaging the entire extension are not mixed with the dependencies of a single component/contribution. 

Here is what im thinking... I like you named the contribution in the manifest terraform-plan (rather than display). So lets carry that name through in the directory structure. Let me know if you think this is reasonable given what you have learned about the UI.

```text
- /tasks
-- /terraform-plan
--- /.bin (compiled source)
--- /.dist (files needed for packaging extension
--- /src (source that exists at the root /src dir moves here)
--- package.json (this is for the dependencies and scripts of the terraform plan extension)
--- tsconfig.json (typescript config for terraform-plan individually)
--- README.md (readme you created under root /src moves here)
```

File: pipelines/install/install.yml [hunk 3]
```diff
@@ -3,11 +3,15 @@ stages:
   dependsOn: build
   variables: 
     - group: env_test
-  jobs: 
+  jobs:
+    # Install extension itself
+    - template: install_terraform_extension.yml
```
Comment (charleszipp) [inline pipelines/install/install.yml:8]: If we install the vsix, do we need the other steps that install the individual tasks? If not lets drop those. An easy way to check is to commit this change and see if the test stage is able to run the tasks.

File: src/plan-summary-tab/plan-summary-tab.tsx [hunk 4]
```diff
@@ -75,59 +77,58 @@ class TerraformPlanDisplay extends React.Component {
...
-                console.log(`Error: ${err}`)
-            })
+    public async componentDidMount() {
+
+        let plan: string | undefined
+        let summary: PlanSummary | undefined
+
+        if (!process.env.TEST) {
+            SDK.init()
+            const build = await this.getThisBuild()
```
Comment (charleszipp) [inline src/plan-summary-tab/plan-summary-tab.tsx:94]: Thanks for updating these to async/await vs the promise xmas tree!

File: src/plan-summary-tab/plan-summary-tab.tsx [hunk 5]
```diff
@@ -75,59 +77,58 @@ class TerraformPlanDisplay extends React.Component {
...
-            })
-            .catch((err) => {
-                console.log(`Error: ${err}`)
-            })
+    public async componentDidMount() {
+
+        let plan: string | undefined
+        let summary: PlanSummary | undefined
+
+        if (!process.env.TEST) {
```
Comment (charleszipp) [inline src/plan-summary-tab/plan-summary-tab.tsx:92]: nit: i would suggest a flip so the first if is positive instead of negative. Its generally easier for people to grok things on positive terms than negative. For example, "if test env var exists then do x" vs. "if test env does not exist then, do x. else do y".

File: vss-extension.json [hunk 6]
```diff
@@ -82,6 +83,21 @@
             "properties": {
                 "name": "TerraformInstaller"
             }
+        },
+        {
+            "description": "A tab to show terraform plan output",
+            "id": "terraform-plan-tab",
```
Comment (charleszipp) [inline vss-extension.json:89]: nit: to be consistent with the others and make sure it doesnt collide with others that might exist in the marketplace, lets rename the id to be `azure-pipelines-tasks-terraform-plan`

File: vss-extension.json [hunk 7]
```diff
@@ -82,6 +83,21 @@
...
+            "type": "ms.vss-build-web.build-results-tab",
+            "targets": [
+                 "ms.vss-build-web.build-results-view"
+            ],
+            "properties": {
+                "name": "Terraform Plan",
+                "supportsTasks": [
```
Comment (charleszipp) [inline vss-extension.json:96]: Why is this needed? I dont recognize that task id.

**Review by charleszipp — CHANGES_REQUESTED** [review]
Looks great @zanawar. Other than the suggested directory reorg, few minor suggestions. Lets talk later this afternoon about the directory structure when you are available. 

Thanks so much for the great work!

File: pipelines/install/install.yml (hunk 3 above)
Comment (zanawar) [reply pipelines/install/install.yml:8]: Nope I don't think we need to do the tasks anymore if we install the vsix, need to confirm this though

File: vss-extension.json (hunk 7 above)
Comment (zanawar) [reply vss-extension.json:96]: I had copied this over from Alex's contribution, i believe that the Task ID needs to be changed to be the "azure-pipelines-tasks-terraform-cli" task now. And doing so will inform ADO that it should show the ui tab if that task is being used.

File: package.json (hunk 2 above)
Comment (zanawar) [reply package.json:6]: This post install might not be required anymore, but I thought it was troubling how running `npm run package:tasks` requires `npm i` to be run in each task (cli and installer), and `npm run pack` requires that `npm package:tasks` to have been run.

I foresaw the developer flow of having to run the following commands just to build to vsix as being unfriendly:
```
npm i
cd tasks/terraform-cli
npm i
npm run build
npm run pack
cd ../terraform-installer
npm i
npm run build
npm run pack
cd ../../
npm run compile
npm run package:tasks
npm run pack
```

Ideally i think the flow should be
```
npm i
npm compile
npm package:tasks
npm pack
```

Perhaps even with the `package:tasks` step combined into `pack`

Comment (charleszipp) [reply package.json]: Thanks for bringing this up. I think what you are suggesting makes alot of sense and tbh I hadnt thought it like that before. It essentially surfaces the question of what does each task/view needs its own package json and to be compiled/tested/etc independently. I am having a hard time coming up with a reason but, may want to think on this a bit more.
Lets discuss when we meet today.

**Review by charleszipp — APPROVED** [review]
